### PR TITLE
TCVP-1285 update data models

### PR DESCRIPTION
**Pull request by karahanb**
# Description

This PR includes the following proposed change(s):

- [TCVP-1285](https://justice.gov.bc.ca/jira/browse/TCVP-1285)
- Refactored existing data model classes, added new ones and updated fields and dependencies.
- Updated the oracle data api dispute service and dispute controller according to the model changes.
- Refactored data models and updated field mapping for message contracts, staff api, arc api and workflow service.
- Changed the disputeDetails and ticketDetails association structure to be based on List and count number instead of array of dictionaries for arc api side.
- Regenerated OpenAPI spec of IOracleDataApi_v1_0Client.json

Here are the Jira tickets that need to be completed after the merge of this PR:

1. TCVP-1302
2.  TCVP-1333
3. TCVP-1334

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Refactoring / Documentation
- [ ] Version change

if your change is a breaking change, please add `breaking change` label to this PR

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes.

## Does the change impact or break the Docker build?

- [ ] Yes
- [x] No

If Yes: Has Docker been updated accordingly?

- [ ] Yes
- [ ] No

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: src/backend/oracle-data-api/src/main/java/ca/bc/gov/open/jag/tco/oracledataapi/model/LegalRepresentation.java
Comment (KDNeufeld) [inline src/backend/oracle-data-api/src/main/java/ca/bc/gov/open/jag/tco/oracledataapi/model/LegalRepresentation.java:45]: We can improve latter on, but I think the typical email regex is more like
"^[\w-\.]+@([\w-]+\.)+[\w-]{2,4}$"
https://regexr.com/3e48o

File: src/backend/oracle-data-api/src/main/java/ca/bc/gov/open/jag/tco/oracledataapi/model/ViolationTicketCount.java
Comment (KDNeufeld) [inline src/backend/oracle-data-api/src/main/java/ca/bc/gov/open/jag/tco/oracledataapi/model/ViolationTicketCount.java:46]: Nice.

**Review by KDNeufeld — APPROVED** [review]
LGTM.  Great work.

File: src/backend/oracle-data-api/src/main/java/ca/bc/gov/open/jag/tco/oracledataapi/model/ViolationTicket.java
Comment (KDNeufeld) [inline src/backend/oracle-data-api/src/main/java/ca/bc/gov/open/jag/tco/oracledataapi/model/ViolationTicket.java:39]: Is there a discriminator or type on here to differentiate between a scanned ViolationTicket and one that is looked up?
Or is the only way to tell is if the ocrViolationTicket json string is empty or not?

Comment (karahanb) [reply src/backend/oracle-data-api/src/main/java/ca/bc/gov/open/jag/tco/oracledataapi/model/ViolationTicket.java:39]: The only way to tell is if the ocrViolationTicket json string is empty or not for now.

File: src/backend/TrafficCourts/Staff.Service/Mappers/Mapper.cs
Comment (KDNeufeld) [inline src/backend/TrafficCourts/Staff.Service/Mappers/Mapper.cs:27]: I didn't realise Section was split out to subsection and paragraph in the model.  I'm not sure it should be.

Here is a listing of the MVA acts we currently have:
https://github.com/bcgov/jag-traffic-courts-online/blob/main/src/backend/oracle-data-api/src/main/resources/data/statutes.csv

This is is what is used to OCR a handwritten ticket (to validate the Section text with the data in the Section column in the csv).  Here are some examples:
```
24.1(1)(b.1)(viii)
24(4) Section (3)(a)
26(1)(b.1)(x)
170(b)(iii)
```
You can see, some of these Section values don't map well to Section, Subsection, and Paragraph used by ARC.

File: src/backend/oracle-data-api/src/main/java/ca/bc/gov/open/jag/tco/oracledataapi/model/ViolationTicketCount.java
Comment (KDNeufeld) [inline src/backend/oracle-data-api/src/main/java/ca/bc/gov/open/jag/tco/oracledataapi/model/ViolationTicketCount.java:89]: Yeah, not sure if we should be splitting out the section/subsection/paragraph from the full Section.
What about the roman numerals often added, ie. (iv)?  The MVA Section data we have from JUSTIN often contains 4 groups, not 3.
24.1(1)(b.1)(viii)
24(4) Section (3)(a)
26(1)(b.1)(x)
170(b)(iii)

Also, I don't know if we should be data fixing JUSTIN data. I kinda think we should just reference and use their data, not try to parse and validate it into separate fields.

File: src/backend/TrafficCourts/Staff.Service/Mappers/Mapper.cs
Comment (karahanb) [reply src/backend/TrafficCourts/Staff.Service/Mappers/Mapper.cs:27]: Yes, I know it's represented as you mentioned in the example on the citizen side and OCR reads it as single section string. However, we have to split it out to section, subsection, paragraph and act fields since ARC file has separate fixed length fields for each so I thought they should be split out at the staff portal level. That's why I added both a "fullSection" field and each field representing other subsections. I think it can be represented on staff portal front-end as separated fields like section, subsection, paragraph and act when it's retrieved from the api from "fullSection" field. So, if staff can verify whether it's accurate and approves the dispute, then it could be mapped to the separated fields on oracle-api and published to the message queue for ARC.

Comment (KDNeufeld) [reply src/backend/TrafficCourts/Staff.Service/Mappers/Mapper.cs:27]: Yeah, I think this needs discussion.  I realize that ARC needs them split out, but I think this logic should be in the ARC message consumer, not in the front-end.  Also, the MVA full Section data is often 4 groups, not three so there is a disparity between what structure ARC is expecting and the structure that is currently in JUSTIN.

Comment (karahanb) [reply src/backend/TrafficCourts/Staff.Service/Mappers/Mapper.cs:27]: In my opinion, it would be safer and more standardized if we can verify the split out data through the staff portal but yes, you're right. I agree that this needs a discussion.

Comment (KDNeufeld) [reply src/backend/TrafficCourts/Staff.Service/Mappers/Mapper.cs:27]: So it sounds like the decision is to have multiple columns in Oracle.  I think we need at least 4 columns for now: section, subsection, paragraph, and clause - maybe a 5th for subclauses later.
From my understanding, for now, ARC will just use the first 3 columns.

**Comment by KDNeufeld** [comment]
I did a `docker-compose up`, unfortunately, the new model in citizen-api confuses swagger, it doesn't load.

http://localhost:5000/swagger/index.html
image

**Comment by pbolduc** [comment]
> I did a `docker-compose up`, unfortunately, the new model in citizen-api confuses swagger, it doesn't load.
> 
> http://localhost:5000/swagger/index.html image

This is cause there are multiple types with the same name.

image

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/ViolationTicket.cs
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Models/Dispute/ViolationTicket.cs:34]: When we lookup a ticket, we dont know this.

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/ViolationTicket.cs
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Models/Dispute/ViolationTicket.cs:33]: When we lookup a ticket, we dont know this.

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/ViolationTicket.cs
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Models/Dispute/ViolationTicket.cs:26]: We already have [TrafficCourts.Citizen.Service.Models.Tickets.ViolationTicketCount](TrafficCourts.Citizen.Service.Models.Tickets.ViolationTicketCount)

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/ViolationTicket.cs
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Models/Dispute/ViolationTicket.cs:7]: We already have [TrafficCourts.Citizen.Service.Models.Tickets.ViolationTicket](https://github.com/bcgov/jag-traffic-courts-online/blob/main/src/backend/TrafficCourts/Citizen.Service/Models/Tickets/ViolationTicket.cs) that has comments

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs:59]: Do not use upper case. What is this Java? Use PascalCase

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs:21]: There is not need to declare the property is IList<T> when it is created as List<T>

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs:7]: The models I have and gave to Shark have comments for all these properties and validation attributes.

File: src/backend/TrafficCourts/Citizen.Service/Mappings/MappingProfile.cs
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Mappings/MappingProfile.cs:9]: There are no unit tests for this mapping profile.

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs:29]: I am still not convinced we should be accepting "ticket" information from the citizen, Either they lookup the ticket or they OCR their ticket. In both cases we cache the data and use that as "the source of truth"

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs:30]: I am wanting to change this `OcrKey` to a ticketId which is created when the user finds or ocr's their ticket.

File: src/backend/TrafficCourts/Citizen.Service/Features/Disputes/Create.cs
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Features/Disputes/Create.cs:83]: I dont think sending email should be here. We should be handling the `SubmitNoticeOfDispute` event in a handler and have it send email

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs:46]: In the screen shots I reviewed, they only want the LawyerFullName not separate fields. Has this changed?

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs:23]: Change to start with upper case.

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs:38]: If the person is pleading not guilty these technically should not be set.

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs
Comment (pbolduc) [inline src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs:39]: If the person is pleading not guilty this technically should not be set.  If pleading not guilty, they have to appear in court.

**Review by pbolduc — CHANGES_REQUESTED** [review]

**Comment by pbolduc** [comment]
We also have to do request validation.

**Comment by karahanb** [comment]
> I did a `docker-compose up`, unfortunately, the new model in citizen-api confuses swagger, it doesn't load.
> 
> http://localhost:5000/swagger/index.html image

The breaking change must have come from the merge I did from main before I push my changes in order to sync with main. It didn't create any conflicts after I merge though. It was working before that.
citizen-api-ss

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/ViolationTicket.cs
Comment (karahanb) [reply src/backend/TrafficCourts/Citizen.Service/Models/Dispute/ViolationTicket.cs:34]: Changed it to use the existing model TrafficCourts.Citizen.Service.Models.Tickets.ViolationTicket

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/ViolationTicket.cs
Comment (karahanb) [reply src/backend/TrafficCourts/Citizen.Service/Models/Dispute/ViolationTicket.cs:33]: Changed it to use the existing model TrafficCourts.Citizen.Service.Models.Tickets.ViolationTicket

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/ViolationTicket.cs
Comment (karahanb) [reply src/backend/TrafficCourts/Citizen.Service/Models/Dispute/ViolationTicket.cs:26]: Changed it to use the existing model TrafficCourts.Citizen.Service.Models.Tickets.ViolationTicketCount

**Review by lfdgov — COMMENTED** [review]
Please check that recent changes pulled into main are reflected in your final PR, at this point it seems to be missing a couple

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs
Comment (karahanb) [reply src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs]: Changed the approach to persist Violation Ticket details and had the looked up tickets to be saved in Redis Cache so the Dispute create request only requires Ticket ID key to be provided to save a dispute along with ticket details.

File: src/backend/TrafficCourts/Citizen.Service/Features/Disputes/Create.cs
Comment (karahanb) [reply src/backend/TrafficCourts/Citizen.Service/Features/Disputes/Create.cs:98]: I haven't made this change and it's already in main so I think it would be better to address this on a separate PR in order to unblock this one.

File: src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs
Comment (karahanb) [reply src/backend/TrafficCourts/Citizen.Service/Models/Dispute/NoticeOfDispute.cs:12]: Added comments and validation attributes to the properties of NoticeOfDispute model in Citizen API.

File: src/backend/TrafficCourts/Citizen.Service/Mappings/MappingProfile.cs
Comment (karahanb) [reply src/backend/TrafficCourts/Citizen.Service/Mappings/MappingProfile.cs:9]: Added unit test for `NoticeOfDispute` to `SubmitNoticeOfDispute` message contract mapping.

File: src/backend/TrafficCourts/Citizen.Service/Features/Disputes/Create.cs
Comment (KDNeufeld) [reply src/backend/TrafficCourts/Citizen.Service/Features/Disputes/Create.cs:98]: Agreed, lets address this in a separate PR.

Comment (lfdgov) [reply src/backend/TrafficCourts/Citizen.Service/Features/Disputes/Create.cs:98]: lets make a jira ticket for this or update an existing one

**Review by lfdgov — APPROVED** [review]

**Comment by karahanb** [comment]
Here are the Jira tickets that need to be completed after the merge of this PR:

1. [TCVP-1302](https://justice.gov.bc.ca/jira/browse/TCVP-1302)
2.  [TCVP-1333](https://justice.gov.bc.ca/jira/browse/TCVP-1333)
3. [TCVP-1334](https://justice.gov.bc.ca/jira/browse/TCVP-1334)

**Review by pbolduc — APPROVED** [review]
Will require completion of linked issues.

**Pull request merged**
